### PR TITLE
feat(hyd): randomization of brake accumulator

### PR DESCRIPTION
**Pull request by crocket63**
Fixes #[issue_no]

## Summary of Changes

This PR brings a realistic starting state for yellow brake accumulator and leaks failures for brakes.

Accumulator might start empty in really rare case (plane cold for 12h, no cargo door operation), in most of cases it will start close to max pressure, according to plane shutdown time.

If empty.... Do not try to remove the chocks.... Just saying....

Accumulator pressure gauge is also adjusted a bit so needle gets to the green band at 3000psi.

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Check brake accumulator gauge is different on various flight restarts.
Check gauge needle goes in the middle of green band when accumulator is fully loaded

Test new brake leaks failure and accumulator gas leak failure.

Check hydraulic quantities in reservoir are ok on multiple restarts and multiple start point (runway/gate...)
Check brakes work as expected in various spawning positions

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by tracernz** [comment]
I don't think we should ever spawn the aircraft in a non-airthworthy state unless the user has selected something to that effect in the flyPad.

**Comment by crocket63** [comment]
> I don't think we should ever spawn the aircraft in a non-airthworthy state unless the user has selected something to that effect in the flyPad.

Nothing non-airworthy here, checking accumulator state and refilling it if needed is part of standard cold start check list.

**Comment by crocket63** [comment]
Last version is broken somehow :(

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 1]
```diff
@@ -263,7 +318,11 @@ impl BrakeCircuit {
     }
 
     pub fn accumulator_fluid_volume(&self) -> Volume {
-        self.accumulator.fluid_volume()
+        if let Some(accumulator) = &self.accumulator {
+            accumulator.fluid_volume()
+        } else {
+            Volume::default()
+        }
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:325]: Could write it like this:
```suggestion
        self.accumulator
            .as_ref()
            .map_or(Volume::default(), |acc| acc.fluid_volume())
```
or like this:
```suggestion
        self.accumulator
            .as_ref()
            .map_or(Volume::default(), Accumulator::fluid_volume)
```

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 2]
```diff
@@ -305,6 +364,24 @@ impl BrakeCircuit {
...
         self.accumulator_fluid_pressure_sensor_filter.output()
     }
+
+    #[cfg(test)]
+    fn accumulator_total_volume(&self) -> Volume {
+        if let Some(accumulator) = &self.accumulator {
+            accumulator.total_volume()
+        } else {
+            Volume::default()
+        }
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:374]:
```suggestion
        self.accumulator
            .as_ref()
            .map_or(Volume::default(), |acc| acc.total_volume())
```

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 3]
```diff
@@ -305,6 +364,24 @@ impl BrakeCircuit {
...
+    }
+
+    #[cfg(test)]
+    fn accumulator_gas_volume(&self) -> Volume {
+        if let Some(accumulator) = &self.accumulator {
+            accumulator.gas_volume()
+        } else {
+            Volume::default()
+        }
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:383]:
```suggestion
       self.accumulator
            .as_ref()
            .map_or(Volume::default(), |acc| acc.gas_volume())
```

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 4]
```diff
@@ -496,6 +583,75 @@ impl Default for AutobrakeDecelerationGovernor {
...
+            random_from_range(0., 1.) < empty_after_maintenance_probability.get::<ratio>();
+
+        let actual_gas_precharge_randomized =
+            Pressure::new::<psi>(random_from_normal_distribution(
+                gas_precharge.get::<psi>(),
+                Self::STANDARD_DEVIATION_FOR_GAS_PRE_CHARGE_DISTRIBUTION_PSI,
+            ));
+
+        let init_volume_for_target_pressure =
+            total_volume - (actual_gas_precharge_randomized * total_volume) / target_pressure;
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:616]: If you change the order of the multiplication and division you get higher precision with the same performance.

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 5]
```diff
@@ -725,22 +881,27 @@ mod tests {
...
 
         assert!(test_bed.query_element(
             |e| e.left_brake_pressure() + e.right_brake_pressure() < Pressure::new::<psi>(10.0)
         ));
 
-        assert!(test_bed.query_element(|e| e.accumulator.total_volume == init_max_vol));
+        assert!(test_bed.query_element(|e| e.accumulator_total_volume() == init_max_vol));
         assert!(
-            test_bed.query_element(|e| e.accumulator.fluid_volume() == Volume::new::<gallon>(0.0))
+            test_bed.query_element(|e| e.accumulator_fluid_volume() == Volume::new::<gallon>(0.0))
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:902]:
```suggestion
            test_bed.query_element(|e| e.accumulator_fluid_volume() == Volume::default())
```

**Review by Gurgel100 — CHANGES_REQUESTED** [review]
Great work as always.

File: src/systems/systems/src/hydraulic/brake_circuit.rs [hunk 6]
```diff
@@ -252,6 +259,54 @@ impl BrakeCircuit {
...
+                            * Self::ACCUMULATOR_GAS_FAILURE_LEAKING_GRADIENT_PSI_PER_S,
+                    ))
+                .max(Pressure::new::<psi>(
+                    Self::ACCUMULATOR_GAS_FAILURE_MIN_ALLOWED_PRESSURE_PSI,
+                ));
+
+                accumulator.set_gas_precharge_pressure(new_pressure_after_leak);
+            } else if self.accu_gas_precharge_failure_active_previous_state
+                && !precharge_failure.is_active()
+            {
```
Comment (Gurgel100) [inline src/systems/systems/src/hydraulic/brake_circuit.rs:288]:
```suggestion
            } else if self.accu_gas_precharge_failure_active_previous_state {
```
The else branch is only executed when the failure is not active so it's not necessary to test that again.

**Review by Gurgel100 — CHANGES_REQUESTED** [review]
Sorry. Somehow this suggestion got lost in my last review.

**Review by Gurgel100 — APPROVED** [review]
LGTM

**Comment by marcman86** [comment]
Quality Assurance Report

Discord: marcman86#4907
Object of testing: #7693
Tier of Testing: 1
Date: 3/02/2023

Testing Process:
- Spawned at gate at EDDP
- Turned on bat. and ext. power
- Checked state of brake accumulator level > about 60%
- Start  yellow epump > brake accu goes green
- Restart
- Checked state of brake accumulator level > about 70%
- Start  yellow epump > brake accu goes green
- Restart
- Checked state of brake accumulator level > about 70%
- Start  yellow epump > brake accu goes green
- Restart over main menu
- Checked state of brake accumulator level > about 70%
- Did this a few more times
- Tested accumulator gas leak failture > brake accu gets empty
- Tested brake leaks failture > yellow circuit fails

Negatives: Randomness works only one time. Stays at level shown in image 1.

Testing Results: Failed
Conclusions: -

Hope this helps :-)

Media:


image 1

**Comment by crocket63** [comment]
> Quality Assurance Report
> 
> Discord: marcman86#4907 Object of testing: #7693 Tier of Testing: 1 Date: 3/02/2023
> 
> Testing Process:
> 
> * Spawned at gate at EDDP
> * Turned on bat. and ext. power
> * Checked state of brake accumulator level > about 60%
> * Start  yellow epump > brake accu goes green
> * Restart
> * Checked state of brake accumulator level > about 70%
> * Start  yellow epump > brake accu goes green
> * Restart
> * Checked state of brake accumulator level > about 70%
> * Start  yellow epump > brake accu goes green
> * Restart over main menu
> * Checked state of brake accumulator level > about 70%
> * Did this a few more times
> * Tested accumulator gas leak failture > brake accu gets empty
> * Tested brake leaks failture > yellow circuit fails
> 
> Negatives: Randomness works only one time. Stays at level shown in image 1.
> 
> Testing Results: Failed Conclusions: -
> 
> Hope this helps :-)
> 
> Media:
> 
>  image 1

What do you exactly mean by randomness only works once?

**Comment by marcman86** [comment]
Test instruction: "Check brake accumulator gauge is different on various flight restarts."

On first spawn gauge shows about 60% (the top one). Then raspawn and gauge shows about 70% like in the image (the top one was the only one who showed a different value). Then respawn again and again 70%. That made me curious. So i tryed a few more respawns. Here it shows every time the same 70%. 

From your test instruction (maybe not precise enough, so i maybe dont understand it the right way) i expeceted the gauge to change on every respawn.

**Comment by crocket63** [comment]
> Test instruction: "Check brake accumulator gauge is different on various flight restarts."
> 
> On first spawn gauge shows about 60% (the top one). Then raspawn and gauge shows about 70% like in the image (the top one was the only one who showed a different value). Then respawn again and again 70%. That made me curious. So i tryed a few more respawns. Here it shows every time the same 70%.
> 
> From your test instruction (maybe not precise enough, so i maybe dont understand it the right way) i expeceted the gauge to change on every respawn.

Looked into this and while the accumulator init is as expected, unwanted brake application on spawn is indeed messing the result.

Could you do another test pass with added checks please?
Check hydraulic quantities in reservoir are ok on multiple restarts and multiple start point (runway/gate...)
Check brakes work as expected in various spawning positions

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -31,6 +31,7 @@
 1. [SOUND] Fix announcements playing twice and adding check for power to PA - @frankkopp (Frank Kopp)
 1. [ENG] Adjust oil pressure table - @tracernz (Mike)
 1. [FBW] Added option to have two axis for rudder e.g. racing pedals - @frankkopp (Frank Kopp)
+1. [HYD] Realistic randomized yellow accumulator starting state - @Crocket63 (crocket)
```
Comment (Gurgel100) [inline .github/CHANGELOG.md:34]: Move to end of changelog

**Comment by 2hwk** [comment]
Quality Assurance Report

Discord: 2Cas#1022
Object of testing: https://github.com/flybywiresim/a32nx/pull/7693
Tier of Testing : 1
Date : 12/02/2023

Testing Process:
-[x] Check brake accumulator gauge is different on various flight restarts. - Took many tries but determined this was working after some bad luck getting similar % a few times in a row. 
-[x] Check gauge needle goes in the middle of green band when accumulator is fully loaded

-[x] Test new brake leaks failure and accumulator gas leak failure.

image
image


-[x] Check hydraulic quantities in reservoir are ok on multiple restarts and multiple start point (runway/gate...)
-[x] Check brakes work as expected in various spawning positions

Negatives:
N/A

Testing Results:
Passed

Conclusions:
LGTM

**Pull request merged**